lib/error_report_notify.c

```diff
@@ -0,0 +1,207 @@
+// SPDX-License-Identifier: GPL-2.0
+
+/*
+ * Userspace notification interface for debugging tools.
+ *
+ * Provide two sysfs files:
+ *  - /sys/kernel/error_report/last_report
+ *  - /sys/kernel/error_report/report_count
+ * that contain the last debugging tool report (taken from dmesg, delimited by
+ * the error_report_start/error_report_end tracing events) and the total report
+ * count.
+ *
+ * Userspace apps can call poll(POLLPRI) on those files to get notified about
+ * the new reports without having to watch dmesg in a loop.
+ *
+ * Because reporting code can be called from low-level routines (e.g. locking
```

Comment:
This is a low-level implementation detail. I'd move this above the function implementing last_report_show, which already mentions some of this. Mentioning it here makes it stand out and it feels like it's very important.

There are also similar low-level implementation details, like the cmpxchg on current_pid, that are of equal importance. These details can just be described above the respective functions that implement the majority of the algorithm.

Reply:
Moved this part of the comment `probe_report_start()`